apps/src/blockly/addons/cdoBlockFlyout.js

```diff
@@ -0,0 +1,212 @@
+      'defs',
+      {},
+      this.svgGroup_
+    );
+    definitions = Blockly.utils.dom.createSvgElement(
+      'clipPath',
+      {
+        id: tagName,
+      },
+      definitions
+    );
+    this.svgClipPath_ = Blockly.utils.dom.createSvgElement(
+      'path',
+      {},
+      definitions
+    );
+    this.svgGroup_.setAttribute('clip-path', 'url(#' + tagName + ')');
+    this.svgGroup_.classList.add('blockFieldFlyout');
+    return this.svgGroup_;
+  }
+
+  /**
+   * Initializes the flyout.
+   *
+   * @param targetWorkspace The workspace in which to
+   *     create new blocks.
+   */
+  init(targetWorkspace) {
+    super.init(targetWorkspace);
+    this.targetWorkspace_ = targetWorkspace;
+  }
+
+  /**
+   * Compute height of flyout.  Position mat under each block.
+   * For RTL: Lay out the blocks right-aligned.
+   */
+  reflowInternal_() {
+    this.height_ = 0;
+    this.width_ = 0;
+    const topBlocks = this.workspace_.getTopBlocks(false);
+    for (let i = 0, block; (block = topBlocks[i]); i++) {
+      const blockHW = block.getHeightWidth();
+      this.updateHeight_(blockHW.height);
+      this.updateWidth_(blockHW.width);
+      if (this.rectMap_.has(block)) {
+        this.moveRectToBlock_(this.rectMap_.get(block), block);
+      }
+    }
+    // We need to set the flyout width based on the block itself, excluding children.
+    // Using block.getHeightWidth() would include blocks connected to input and next
+    // connections, which could make the flyout wider than the block that contains it.
+    const topBlockWidth = this.sourceBlock_.svgGroup_
+      .querySelector('path')
+      .getBoundingClientRect().width;
+    this.width_ = Math.max(this.width_, topBlockWidth - 36);
+    this.setBackgroundPath_(this.width_, this.height_);
+    this.position();
+  }
+
+  /** Update the flyout height based on the new block height.
+   *
+   * @param {number} newHeight - The new block height.
+   * @private
+   */
+  updateHeight_(newHeight) {
+    this.height_ = Math.max(this.height_, newHeight);
+  }
+  /**
+   * Update the flyout width based on the new block width.
+   *
+   * @param {number} newWidth - The new block width.
+   * @private
+   */
+  updateWidth_(newWidth) {
+    this.width_ += newWidth + this.GAP_X;
+  }
+
+  /**
+   * Position the flyout.
+   */
+  position() {
+    this.isVisible() && this.positionAt_(this.width_, this.height_, 0, 0);
+  }
+
+  /**
+   * Create and set the path for the visible boundaries of the flyout.
+   *
+   * @param width The width of the flyout, not including the rounded corners.
+   * @param height The height of the flyout, not including rounded corners.
+   */
+  setBackgroundPath_(width, height) {
+    const svgPaths = GoogleBlockly.utils.svgPaths;
+    let path = svgPaths.moveTo(
+      this.RTL ? width : 0,
+      this.toolboxPosition_ === Blockly.TOOLBOX_AT_TOP ? 0 : this.CORNER_RADIUS
+    );
+
+    path += svgPaths.curve('a', [
```

Comment:
is there a way we can extract out this logic (lines 141-148, repeated a few times below) into some sort of helper function? It looks like the only difference is setting positive and negative values?

Reply:
Totally reworked, PTAL!